### ISSUE TITLE
Clipboard vulnerability; replace everywhere by Drag & Drop

### ISSUE THREAD

**Issue by W-H-H** [keepassxreboot/keepassxc#2724]
-------
System
- Linux - Xubuntu 18.04 LTS (32-bit), fully up to date and fully patched
- KeePassXC 2.3.4-1ppa1~bionic1
- Firefox 65.0
- KeePassXC-Browser 1.3.2 (official Firefox browser add-on (browser integration plug-in))
-------

KeePassXC makes extensive use of the option to copy usernames and decrypted passwords to the operating system's clipboard. On the other hand, KeePassXC does not properly support drag & drop of plaintext usernames and passwords into and out of the KeePassXC environment.

This is a major security vulnerability and a hazard for KeePassXC users for at least two reasons:

(1) System clipboards store text as user-invisible but directly accessible plaintext. This can inadvertently find its way into other documents or into wrong web page entry fields, for example through an accidental Ctrl+V or through entry field misinterpretation by KeePassXC itself.

Visually-impaired users are just one category of users typically and immediately at risk here.

(2) KeePassXC makes an attempt somewhat to reduce this risk by allowing the user (via its Settings dialog) to set a (short) preservation time limit on the clipboard contents.

However, this mitigation is entirely ineffective, for example, when the user has installed a clipboard manager.

There are numerous good reasons for using a clipboard manager. For example, it transparently displays the (otherwise invisible) contents of the clipboard. It allows for the efficient recycling of standard words and phrases. And for the disabled, it can be a valuable accessibility tool.

(2.1) Clipboard managers override the time limit which KeePassXC tries to enforce. Therefore, decrypted passwords sent to the clipboard using a copy instruction in KeepassXC are preserved indefinitely.

(2.2) And these decrypted passwords are preserved in plaintext, written to an unprotected clipboard manager file on the hard drive that is open to any intruder and malware.

This completely defeats the very purpose itself to securely encrypt the KeePassXC password database.

In short, KeePassXC makes itself dependent on operating system functionality, namely the clipboard, over which it has no sole and exclusive control. This is a fundamental security flaw.

The only way to eliminate this security risk is for KeePassXC to remove and eliminate any and all options to use the operating system's copy-to-clipboard and paste-from-clipboard functions.

And the only acceptably secure alternative is entirely to replace the copy-to-clipboard options, wherever they now occur in the operational KeePassXC environment, by comprehensive drag & drop functionality.

Drag & drop is not normally accessible to any subsequent user operation. It leaves no accessible record on the hard drive (beyond potentially an encrypted swap file, if used). And any remaining trace is erased from memory on system shut-down.

Drag & drop is as secure as, for example, displaying decrypted passwords when using KeePassXC's "View/Edit entry" function.

**Comment by phoerious** [comment]
> This completely defeats the very purpose itself to securely encrypt the KeePassXC password database.

It does not. Yes, the clipboard is not the most secure channel, but a convenient one. Certain clipboard managers such as KDE's Klipper support setting flags to not remember certain clipboard contents (which we use). 
If you are worried about programs reading your clipboard, use Auto-Type or the browser extension.

**Comment by droidmonkey** [comment]
I do not agree with your assertions on the security of Drag and Drop. It is only as secure as OLE is on Windows, XDND is on Linux, and Cocoa Drag Manager is on MacOS.

https://doc.qt.io/qt-5/dnd.html#interoperating-with-other-applications

I doubt greatly that any of these implementations has "password sensitive" security in mind when they were implemented.

**Comment by W-H-H** [comment]
> > This completely defeats the very purpose itself to securely encrypt the KeePassXC password database.
> 
> It does not. Yes, the clipboard is not the most secure channel, but a convenient one. Certain clipboard managers such as KDE's Klipper support setting flags to not remember certain clipboard contents (which we use).
> If you are worried about programs reading your clipboard, use Auto-Type or the browser extension.

Klipper is a KDE program. But KDE is not the most extensively used Linux desktop environment by a long shot (and it is resource intensive at that).

Specifically, klipper is available only if the full KDE desktop environment is installed; it is, for example, not available as an individually-installable application in the Ubuntu repositories, the most widely used Linux distribution (nor, for that matter, is klipper's GNOME equivalent, glipper). So this is not a solution to the security vulnerability identified.

Further and more importantly, _it is up to KeePassXC to aim and ensure that it is always secure_.

Making itself dependent on an insecure general-purpose user interface facility, the operating system's clipboard, a feature which was never designed as secure and which is patently and inherently insecure, is wrong by design.

The convenience which you cite is no excuse. It reflects a cavalier attitude to security.

If KeePassXC really needs copy & paste functionality, _then it should use **its own secure clipboard implementation**_.

As alternatives to clipboard use, you mention auto-type and the KeePassXC browser extension.

Auto-type is functional, but only if the web site's login fields are correctly identified. Misidentification of the latter currently requires the browser extension for correction.

At present, the KeePassXC browser extension's ability correctly to identify login fields breaks down on many web sites due to their frame use and scripting. In addition, the browser extension's context menu for saving login credentials to the KeePassXC database and for entering saved login credentials from the KeePassXC database is not functional (see #2725).

The key point is this:

As a privacy and security application, KeePassXC goes to very great lengths to secure the password database through state-of-the-art encryption. That is fine, it should do that.

But it shirks its responsibility if then, in actual operational use, it is happy to rely on inherently insecure data transmission channels between itself and its target, that is, the login page of a web site.

 _Secure means only that if it is **end-to end**_.

**Comment by W-H-H** [comment]
@droidmonkey 

> I do not agree with your assertions on the security of Drag and Drop. It is only as secure as OLE is on Windows, XDND is on Linux, and Cocoa Drag Manager is on MacOS.
> 
> https://doc.qt.io/qt-5/dnd.html#interoperating-with-other-applications
> 
> I doubt greatly that any of these implementations has "password sensitive" security in mind when they were implemented. You can freely choose not to use the clipboard by simply not copy and pasting information from your database. There is no reason to remove the functionality, perhaps a setting to disable it would be acceptable but not a priority at this time.


As set out in my original submission on this issue, the general-purpose clipboard _is_ significantly less secure than drag & drop, and in actual operational practice especially so for visually impaired users and for users of many (but -- if properly set up with security and transparency in mind -- apparently not all) clipboard managers.

Further, even in the absence of a clipboard manager, a clipboard cut/copy event always copies login credentials expressly in a _non-volatile_ manner, fire and forget, in plaintext freely accessible by design to all comers. Both preservation and unhindered availability to unspecified other clients are deliberate; intentionally there is _not a defined, let alone an exclusively defined, target client_.

In most implementations the current clipboard contents (that is, the login credentials unless subsequently overwritten by another unrelated cut/copy event) are also copied in plaintext to disk on system shutdown for access in subsequent sessions. This hardens permanence and unhindered accessibility yet more.

Drag & drop is at least volatile always.


In addition, for example, a keystroke logger or other malware has every opportunity to scrape the clipboard at its leisure.

A _user-set time limit_ after which KeePassXC clears the clipboard is mere _ineffective sticking plaster_  here: A user who employs copy & paste for login credentials submission requires several seconds for the clipboard contents to remain available, but a clipboard scraper which is actively monitoring the clipboard requires only microseconds if not less.


In contrast, it requires considerably more sophistication reliably to intercept drag & drop actions:

Specifically, drag & drop events require _explicit signalling between the source and the identified target_. The unique legitimate target for any individual login event is known to KeePassXC. Therefore, KeePassXC can easily prevent both erroneous misdirection and malicious interception of login credentials under drag & drop.

In fact, if KeePassXC does not put adequate privacy and security constraints on such signalling -- which, from your comments, seems to be the case --, _then it needs to implement those as a matter of urgency_.

Simply relying on system standards which have never been designed with privacy and security in mind, as you apparently suggest, amounts to a fundamental abandonment of purpose on the side of KeePassXC.


As things stand, the current version of the KeePassXC browser extension is defective and auto-type often fails (see #2725).

Therefore, at present the user is effectively forced to use fundamentally insecure copy & paste actions from the main database window of the KeePassXC desktop application: The latter does not allow the inherently more secure drag & drop. As an avoidable security vulnerability, this needs to be addressed.


I would, albeit most reluctantly, respect a decision not to eliminate copy & paste functionality, but only for those users who simply cannot memorize numerous different passwords and who consciously are not terribly interested in password security (although they might as well simply store their passwords in a plaintext file rather than in a KeePassXC database).

However, for users who are concerned about privacy and security, there should then at the very least be a general settings option to remove all cut, copy & paste actions via the default operating system clipboard from the main and context menus everywhere in KeePassXC _so as to avoid inadvertent use_ and _so as to avoid the suggestion that it is safe and transparent to use_.

They should be offered instead a **secure KeePassXC implementation of the clipboard** and **full and secured drag & drop functionality**.


In summary, as things stand given the absence of drag & drop functionality and given a dysfunctional browser extension, it is currently safer (in the absence of shoulder surfers) to:

 (1) enter all web site login credentials with mousepad, gedit or similar in a plain text file
 (2) securely encrypt this text file with a master password using ubiquitous state-of-the-art file encryption software
 (3) unlock this text file when a web site login is required
 (4) drag & drop the login credentials from the unlocked text file

I cannot imagine that KeePassXC would wish to see itself defeated on privacy and security grounds by such a simple device.

**Comment by phoerious** [comment]
> I would, albeit most reluctantly, respect a decision not to eliminate copy & paste functionality through the operating system clipboard, but only for those users who simply cannot memorize numerous different passwords and who consciously are not terribly interested in password security (although they might as well simply store their passwords in a plaintext file rather than in a KeePassXC database).

This statement is simply false. KeePassXC is always more secure than a plaintext file opened with an ordinary text editor. We do not save temporary unencrypted copies somewhere, we at least try to take care of clearing the clipboard, we do not automatically store a selected entry in the X11 primary clipboard (a text editor will do that as soon as you select your password from the file in order to drag it somewhere), and we employ various mitigations for preventing unauthorized memory access.

On the other hand, to exploit the overall lower security of the clipboard, you need a local attacker on your system, in which case clipboard snooping is certainly one, but probably not the highest of your concerns. You be more worried about that attacker installing fake root certificates and recording your browser session. Once your system has been compromised with malware, there is no effective protection against data theft, no matter if you type, auto-type, copy/paste, or drag/drop your passwords. All we can do is try to mitigate the damage.

We are open for the suggestion of better supporting drag and drop of credentials. But please stop suggesting that KeePassXC is inherently insecure in lack of it and not using a password manager would be more secure, because that could not be farther from the truth.

**Comment by ChristopherSchultz** [comment]
Whether or not copy/paste should be purged from keepassxc, I really do miss having drag-and-drop password capability on MacOS. It's literally the only thing about keepassx that I really miss. But I *really* do miss it.

Please consider dedicating some resources to issue #1987.

**Comment by ChristopherSchultz** [comment]
> Please consider dedicating some resources to issue #1987.

... because auto-type doesn't really work with CLI tools. `sudo make me a sandwich`, then `Password:[?]` :( Auto-type for window name `Terminal` isn't going to cut it.

**Comment by droidmonkey** [comment]
Ctrl+Shift+V works fantastic for me in this circumstance

**Comment by martinpelikan** [comment]
Just started looking into using a clipboard manager recently and found [CopyQ](https://hluk.github.io/CopyQ/) which allows you to [omit copying from specific windows](https://copyq.readthedocs.io/en/latest/faq.html#how-to-omit-storing-text-copied-from-specific-windows-like-a-password-manager). It also has settings to encrypt the clipboard history or avoid saving it altogether. 

I can't speak for how secure this is, but it appears to be an improvement over using a plain clipboard manager that will save the password indefinitely.
